File: Assets/Framework/ShibaInu/Utils/Logging/Logger.cs
using System;
using UnityEngine;
using LuaInterface;


namespace ShibaInu
{

    /// <summary>
    /// 日志记录
    /// </summary>
    public static class Logger
    {
        private static LuaFunction s_uncaughtExceptionHandler;


        [NoToLua]
        public static void Initialize()
        {
            Application.logMessageReceived += UnityLogCallback;
        }


        /// <summary>
        /// Unity 日志回调
        /// </summary>
        /// <param name="condition">Condition.</param>
        /// <param name="stackTrace">Stack trace.</param>
        /// <param name="type">Type.</param>
        private static void UnityLogCallback(string condition, string stackTrace, LogType type)
        {
            switch (type)
            {
                case LogType.Exception:
                case LogType.Assert:
                    string logType = type.ToString();
                    stackTrace = string.Format("\nstack traceback:\n\t{0}", stackTrace.TrimEnd().Replace("\n", "\n\t"));
                    LogData.Append(logType, condition, stackTrace);
                    UncaughtExceptionHandler(logType, condition, stackTrace);
                    break;

                case LogType.Warning:
                    // 将 DoTween 回调中的报错从 Warning 转为 Error
                    int idx = condition.IndexOf("An error inside a tween callback");
                    if (idx != -1)
                    {
                        condition = condition.Substring(condition.IndexOf('►', idx) + 1);
                        int idx1 = condition.IndexOf('\n');
                        int idx2 = condition.IndexOf("\n\n", idx1);
                        stackTrace = condition.Substring(idx1, idx2 - idx1);
                        condition = condition.Substring(0, idx1);
                        condition = "[tween callback]" + condition;
                        LogError(condition, stackTrace);
                        UncaughtExceptionHandler(LogData.TYPE_ERROR, condition, stackTrace);
                    }
                    break;
            }
        }


        /// <summary>
        /// 调用异常收集回调
        /// </summary>
        /// <param name="type"></param>
        /// <param name="msg"></param>
        /// <param name="stackTrace"></param>
        private static void UncaughtExceptionHandler(string type, string msg, string stackTrace)
        {
            if (s_uncaughtExceptionHandler != null)
            {
                s_uncaughtExceptionHandler.BeginPCall();
                s_uncaughtExceptionHandler.Push(type);
                s_uncaughtExceptionHandler.Push(msg);
                s_uncaughtExceptionHandler.Push(stackTrace);
                s_uncaughtExceptionHandler.PCall();
                s_uncaughtExceptionHandler.EndPCall();
            }
        }


        /// <summary>
        /// 设置异常收集回调
        /// </summary>
        /// <param name="callback">Callback.</param>
        public static void SetUncaughtExceptionHandler(LuaFunction callback)
        {
            s_uncaughtExceptionHandler = callback;
        }



        /// <summary>
        /// 添加一条普通日志
        /// </summary>
        /// <param name="msg">Message.</param>
        /// <param name="type">Type.</param>
        /// <param name="stackTrace">Stack trace.</param>
        public static void Log(string msg, string type = LogData.TYPE_LOG, string stackTrace = null)
        {
            LogData data = LogData.Append(type.Trim(), msg, stackTrace);
            Debug.Log(data);
        }


        /// <summary>
        /// 添加一条警告日志
        /// </summary>
        /// <param name="msg">Message.</param>
        /// <param name="stackTrace">Stack trace.</param>
        public static void LogWarning(string msg, string stackTrace)
        {
            LogData data = LogData.Append(LogData.TYPE_WARNING, msg, stackTrace);
            Debug.LogWarning(data);
        }


        /// <summary>
        /// 添加一条错误日志
        /// </summary>
        /// <param name="msg">Message.</param>
        /// <param name="stackTrace">Stack trace.</param>
        public static void LogError(string msg, string stackTrace)
        {
            LogData data = LogData.Append(LogData.TYPE_ERROR, msg, stackTrace);
            Debug.LogError(data);
        }


        /// <summary>
        /// 添加一条网络日志
        /// </summary>
        /// <param name="msg">Message.</param>
        /// <param name="type">Type.</param>
        /// <param name="info">Info.</param>
        public static void LogNet(string msg, string type, string info)
        {
            LogData.Append(type, msg, "\ninfo: " + info);
        }



        /// <summary>
        /// [C#] 添加一条异常日志
        /// </summary>
        /// <param name="exception">Exception.</param>
        [NoToLua]
        public static void LogException(Exception exception)
        {
            Debug.LogException(exception);
        }

        [NoToLua]
        public static void LogException(string message)
        {
            Debug.LogException(new Exception(message));
        }



        #region 清空所有引用（在动更结束后重启 app 时）

        [NoToLua]
        public static void ClearReference()
        {
            s_uncaughtExceptionHandler = null;
        }

        #endregion


        //
    }
}

File: Assets/Framework/ShibaInu/Lua/Utils/Logging/Logger.lua
--
-- 日志
-- 2017/10/18
-- Author LOLO
--


local xpcall = xpcall
local traceback = debug.traceback
local unpack = unpack or table.unpack

local log = ShibaInu.Logger.Log
local logWarning = ShibaInu.Logger.LogWarning
local logError = ShibaInu.Logger.LogError
local logNet = ShibaInu.Logger.LogNet

local TYPE_TRACE = "Trace" -- 默认日志
local TYPE_ERROR = "Error" -- 报错日志
local TYPE_NET_SUCC = "NetSucc" -- 网络日志 - 通信成功
local TYPE_NET_FAIL = "NetFail" -- 网络日志 - 通信失败
local TYPE_NET_PUSH = "NetPush" -- 网络日志 - 主动推送


--
---@class Logger
local Logger = {}

---@type fun(type:string, msg:string, stackTrace:string)
local uncaughtExceptionHandler



--
--- 添加一条普通日志
---@param msg string @ 日志内容
---@param type string @ -可选- 日志类型。默认："Trace"
---@param isStackTrace boolean @ -可选- 是否记录堆栈信息。默认：true
function Logger.Log(msg, type, isStackTrace)
    type = type or TYPE_TRACE

    if isStackTrace ~= false then
        log(msg, type, traceback("", 2)) -- 包含堆栈信息
    else
        log(msg, type)
    end
end


--
--- 添加一条警告日志
---@param msg string @ 日志内容
function Logger.LogWarning(msg)
    logWarning(msg, traceback("", 2))
end


--
--- 添加一条错误日志
---[Error Handler]
---@param msg string @ 日志内容
function Logger.LogError(msg)
    local stackTrace = traceback("", 2)
    logError(msg, stackTrace)
    if uncaughtExceptionHandler ~= nil then
        uncaughtExceptionHandler(TYPE_ERROR, msg, stackTrace)
    end
end




--
--- 添加一条网络日志，通信成功
---@param msg string @ 标题
---@param info string @ 详情
function Logger.LogNetSucc(msg, info)
    logNet(msg, TYPE_NET_SUCC, info)
end

--
--- 添加一条网络日志，通信失败
---@param msg string @ 标题
---@param info string @ 详情
function Logger.LogNetFail(msg, info)
    logNet(msg, TYPE_NET_FAIL, info)
end

--
--- 添加一条网络日志，主动推送
---@param msg string @ 标题
---@param info string @ 详情
function Logger.LogNetPush(msg, info)
    logNet(msg, TYPE_NET_PUSH, info)
end



-- try call
local errorHandler = Logger.LogError

-- 确定是否可以使用新版本（lua5.2 or newer）的 xpcall
-- 使用新版本的 xpcall 可以减少一次匿名函数封装和 unpack() 调用
local isLatestXpcall
xpcall(function(isLatest)
    isLatestXpcall = isLatest == true
end, function()
end, true)

--
--- 调用 fn，并捕获出现的错误（ try ... catch ）
---@param fn fun() @ 传入的函数
---@param caller any @ -可选- self 对象，默认为 nil
---@vararg any @ -可选- 附带的参数
---@return boolean, any @ 调用函数是否成功（是否没有报错），以及 fn 函数返回值
if isLatestXpcall then
    Logger.TryCall = function(fn, caller, ...)
        if caller == nil then
            return xpcall(fn, errorHandler, ...)
        else
            return xpcall(fn, errorHandler, caller, ...)
        end
    end
else
    Logger.TryCall = function(fn, caller, ...)
        local args = { ... }
        if caller == nil then
            return xpcall(function()
                fn(unpack(args))
            end, errorHandler)
        else
            return xpcall(function()
                fn(caller, unpack(args))
            end, errorHandler)
        end
    end
end



--
--- 设置 Lua 或 C# 出现未捕获异常时的回调
---@param callback fun(type:string, msg:string, stackTrace:string) @ 报错时的回调，用于收集报错日志等
function Logger.SetUncaughtExceptionHandler(callback)
    uncaughtExceptionHandler = callback
    ShibaInu.Logger.SetUncaughtExceptionHandler(callback)
end



--
return Logger
